### ISSUE TITLE
views/authorization_rules/index.html.erb where line #16 raised: "one hash required"

### ISSUE THREAD

**Issue by rodrigob** [stffn/declarative_authorization#30]
When looking at /authorization_rules I get the folllowing error:

  Showing /var/lib/gems/1.8/gems/declarative_authorization0.4/app/views/authorization_rules/index.html.erb where line #16 raised:

  one hash required
  Extracted source (around line #16):

  13:   pre.with-notes {padding-left: 35px}
  14: </style>
  15: <pre class="with-notes">
  16: <%=   policy_analysis_hints(syntax_highlight(h(@auth_rules_script)),@auth_rules_script) %>
  17: </pre>
  RAILS_ROOT: /home/rodrigob/work/popster_work/popster_bzr/src/web/popster

  Application Trace | Framework Trace | Full Trace
  /var/lib/gems/1.8/gems/i18n-0.3.3/lib/i18n/core_ext/string/interpolate.rb:88:in `%'
  /var/lib/gems/1.8/gems/declarative_authorization-0.4/lib/declarative_authorization/development_support/analyzer.rb:107:in`message'
  /var/lib/gems/1.8/gems/declarative_authorization-0.4/lib/declarative_authorization/development_support/analyzer.rb:89:in `mark'
  /var/lib/gems/1.8/gems/declarative_authorization-0.4/lib/declarative_authorization/development_support/analyzer.rb:49:in`analyze'
  /var/lib/gems/1.8/gems/declarative_authorization-0.4/lib/declarative_authorization/development_support/analyzer.rb:35:in `analyze'
  /var/lib/gems/1.8/gems/declarative_authorization-0.4/lib/declarative_authorization/development_support/analyzer.rb:32:in`each'
  /var/lib/gems/1.8/gems/declarative_authorization-0.4/lib/declarative_authorization/development_support/analyzer.rb:32:in `analyze'
  /var/lib/gems/1.8/gems/declarative_authorization-0.4/app/helpers/authorization_rules_helper.rb:27:in`policy_analysis_hints'
blahblahblah....

created a dummy fix by changing lines 106-107-108 of
declarative_authorization-0.4/lib/declarative_authorization/development_support/analyzer.rb
into

```
    def message (object)
      "The ratio of small roles is quite high (> %.0f).  Consider refactoring." % (SMALL_ROLES_RATIO * 100)
    end
```

most certainly not the correct fix, but I'm a total newbie with ruby and rails.

**Comment by stffn** [comment]
Which Ruby version are you using?  I don't have any problems with that line, even when run in irb.

**Comment by rodrigob** [comment]
$ ruby --version
ruby 1.8.7 (2009-06-12 patchlevel 174) [i486-linux]

(ubuntu 9.10)

rails  2.3.5

**Comment by stffn** [comment]
This is odd.  I can't reproduce the error on the same Ruby version.  Did you change the constant SMALL_ROLES_RATIO?  Does the original statement work on your irb console?

  "The ratio of small roles is quite high (> %.0f%%).  Consider refactoring." % (0.2 \* 100)

**Comment by rodrigob** [comment]
Well my fix was to use
"The ratio of small roles is quite high (> %.0f).  Consider refactoring." % (SMALL_ROLES_RATIO \* 100)

so I do not think the problem as SMALL_ROLES_RATIO.
It seemed that the output string
"blahblah (20%) blohbloh" was reinterpreted again and the '%' was processed as an input field in the string, hence the "one hash required" error.
It seems to be expecting "blahblah (20%) blohbloh" % {some_data}

**Comment by QuBiT** [comment]
I got the same Error as Rodrigob:

<pre>

"one hash required"

On line #16 of /usr/lib/ruby/gems/1.8/gems/declarative_authorization-0.4.1/app/views/authorization_rules/index.html.erb

13: pre.with-notes {padding-left: 35px}
14: </style>
15: <_pre class="with-notes">
16: <%= policy_analysis_hints(syntax_highlight(h(@auth_rules_script)), @auth_rules_script) %>
17: </_pre>

/usr/lib/ruby/gems/1.8/gems/i18n-0.3.3/lib/i18n/core_ext/string/interpolate.rb:88:in `%'
/usr/lib/ruby/gems/1.8/gems/declarative_authorization-0.4.1/app/lib/declarative_authorization/development_support/analyzer.rb:107:in `message'
/usr/lib/ruby/gems/1.8/gems/declarative_authorization-0.4.1/app/lib/declarative_authorization/development_support/analyzer.rb:89:in `mark'
/usr/lib/ruby/gems/1.8/gems/declarative_authorization-0.4.1/app/lib/declarative_authorization/development_support/analyzer.rb:49:in `analyze'
/usr/lib/ruby/gems/1.8/gems/declarative_authorization-0.4.1/app/lib/declarative_authorization/development_support/analyzer.rb:35:in `analyze'
/usr/lib/ruby/gems/1.8/gems/declarative_authorization-0.4.1/app/lib/declarative_authorization/development_support/analyzer.rb:32:in `each'
/usr/lib/ruby/gems/1.8/gems/declarative_authorization-0.4.1/app/lib/declarative_authorization/development_support/analyzer.rb:32:in `analyze'
/usr/lib/ruby/gems/1.8/gems/declarative_authorization-0.4.1/app/helpers/authorization_rules_helper.rb:27:in `policy_analysis_hints'
/usr/lib/ruby/gems/1.8/gems/declarative_authorization-0.4.1/app/views/authorization_rules/index.html.erb:16 

</pre>


same system as rodrigob (ubuntu 9.10, rails 2.3.5, ...) using declarative_authorization-0.4.1

**Comment by rodrigob** [comment]
The problem seems related to the i18n gem
http://groups.google.com/group/rails-i18n/browse_thread/thread/9232cc9e154002ea/a6785bc23982b16c

yet my application needs i18n support...

**Comment by stffn** [comment]
If this isn't a decl_auth error, you have two choices:  monkey-patch the decl_auth method (as a workaround) or monkey-patch i18n as a fix (and submit a patch to i18n).

**Comment by rodrigob** [comment]
Ok, I could not figure out.
i18n was installed when I installed rails 3.0.0.beta and for some strange reason that I do not understand, installing i18n messed up my rails 2.3.5 applications.

Since I was just toying with rails3, I removed i18n from my system, and now "everything is fine". Such "across gems" effects are a really uncool. 

I cannot tell if i18n has a nasty bug or decl_auth is not rails3 and rails2 compatible, or both.
